### ISSUE TITLE
Improved agent boot times

### ISSUE THREAD

**Issue by Stono** [gocd/gocd#5061]
Hey, 
We're using kubernetes ephemeral agents with gocd, which is great to ensure we have deterministic builds, however the additional start time of creating a new pod and thus registering a new elastic agent is around 40-50s (mainly agent boot + register time).

I'm wondering if you have any ideas of ways we could improve the boot time for these elastic agents?

**Comment by ketan** [comment]
Hi there —

While we don't have have an official solution to this — the GoCD team has been using [this](https://github.com/ketan/gocd-golang-bootstrapper/) for quite some time now. It's fully compatible with the k8s/docker/swarm/ecs plugins that we have so far. The startup speeds are fairly quick < 2-3 seconds, and a lower footprint. You'd normally use it as:

```Dockerfile
FROM ...

# The directory and user that'll be used for the gocd agent process. Currently hardcoded in the bootstrapper ;-/
RUN useradd --home-dir /go --shell /bin/bash go

# Download and ensure that the binary is executable
RUN curl --fail --silent --location https://github.com/ketan/gocd-golang-bootstrapper/releases/download/2.1/go-bootstrapper-2.1.linux.amd64 > /go-agent && \
  chmod 755 /go-agent

# execute the binary, preferably as the `go` user
USER go
e /go-agent
```

**Comment by Stono** [comment]
If this works this is awesome!  Thanks

**Comment by Evesy** [comment]
Hey @ketan 

Trying this out the agent is successfully booted & I can see it has registered itself with the GoCD server, however it seems to then hang indefinitely and is never assigned the job that it has been created for:

```
2018-08-28T11:05:23.709Z+00:00 [Bootstrap] [DEBU] Getting checksums from https://goqa.atcloud.io/go/admin/latest-agent.status
2018-08-28T11:05:23.893Z+00:00 [Bootstrap] [DEBU] agent.jar                     - 5e51a7cb2e55fa232e8a1d25e6e2aba7
2018-08-28T11:05:23.893Z+00:00 [Bootstrap] [DEBU] agent-plugins.zip             - 47383b4ad0fb2c52313aad5e77b18a03
2018-08-28T11:05:23.893Z+00:00 [Bootstrap] [DEBU] tfs-impl.jar                  - fa884c713ea0dc9b1430dc551155124f
2018-08-28T11:05:23.894Z+00:00 [Bootstrap] [DEBU] agent-launcher.jar (not used) - 9c67d216e09d2b82a17c8fcfd3bd48c5
2018-08-28T11:05:23.894Z+00:00 [Bootstrap] [INFO] Downloading file agent.jar from https://goqa.atcloud.io/go/admin/agent
2018-08-28T11:05:29.922Z+00:00 [Bootstrap] [INFO] Finished downloading file agent.jar from https://goqa.atcloud.io/go/admin/agent
2018-08-28T11:05:29.922Z+00:00 [Bootstrap] [INFO] Downloading file agent-plugins.zip from https://goqa.atcloud.io/go/admin/agent-plugins.zip
2018-08-28T11:05:36.206Z+00:00 [Bootstrap] [INFO] Finished downloading file agent-plugins.zip from https://goqa.atcloud.io/go/admin/agent-plugins.zip
2018-08-28T11:05:36.206Z+00:00 [Bootstrap] [INFO] Downloading file tfs-impl.jar from https://goqa.atcloud.io/go/admin/tfs-impl.jar
2018-08-28T11:05:40.745Z+00:00 [Bootstrap] [INFO] Finished downloading file tfs-impl.jar from https://goqa.atcloud.io/go/admin/tfs-impl.jar
2018-08-28T11:05:40.745Z+00:00 [Bootstrap] [INFO] Launching command [java -Dcruise.console.publish.interval=10 -Xms128m -Xmx256m -Djava.security.egd=file:/dev/./urandom -Dagent.plugins.md5=47383b4ad0fb2c52313aad5e77b18a03 -Dagent.binary.md5=5e51a7cb2e55fa232e8a1d25e6e2aba7 -Dagent.launcher.md5=9c67d216e09d2b82a17c8fcfd3bd48c5 -Dagent.tfs.md5=fa884c713ea0dc9b1430dc551155124f -jar agent.jar -serverUrl https://goqa.atcloud.io/go -sslVerificationMode FULL] with environment [PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin HOSTNAME=gocd-admin-agent-5e226ebe-55c8-4142-a503-a04e7551222c GOOGLE_APPLICATION_CREDENTIALS=/etc/gcloud/gcloud_json GO_SERVER_API_URL=https://goqa.atcloud.io/go/api GOCD_AGENT_PULL_BASE=false DOCKER_HOST=tcp://127.0.0.1:2375 TEST_ENV="true" GO_API_PASSWORD=<REDACTED> KUBERNETES_PORT_443_TCP=tcp://10.192.104.1:443 KUBERNETES_PORT_443_TCP_PROTO=tcp KUBERNETES_PORT_443_TCP_PORT=443 KUBERNETES_PORT_443_TCP_ADDR=10.192.104.1 KUBERNETES_SERVICE_HOST=10.192.104.1 KUBERNETES_SERVICE_PORT=443 KUBERNETES_SERVICE_PORT_HTTPS=443 KUBERNETES_PORT=tcp://10.192.104.1:443 APP_DIR=/usr/local/autotrader/app TERM=xterm-256color JAVA_HOME=/usr/lib/jvm/java-openjdk APP_PORT=8080 APP_ADMIN_PORT=9080 JMX_METRICS_PORT=9090 HEAP_DUMP_DIR=/atcloud/heapdumps APP_PATH=/usr/local/autotrader/app HOME=/go]
2018-08-28T11:05:40.746Z+00:00 [Bootstrap] [INFO] Agent process launched PID(28), waiting for it to exit.
[Tue Aug 28 11:05:41 UTC 2018] Starting process:
[Tue Aug 28 11:05:41 UTC 2018]   Working directory    : /go
[Tue Aug 28 11:05:41 UTC 2018]   Application arguments: [-serverUrl, https://goqa.atcloud.io/go, -sslVerificationMode, FULL]
[Tue Aug 28 11:05:41 UTC 2018]            GoCD Version: 18.6.0-6883
[Tue Aug 28 11:05:41 UTC 2018]   JVM arguments        : [-javaagent:/usr/local/bin/jmx_prometheus_javaagent.jar=9090:/usr/local/bin/javaagent.yaml, -Dcruise.console.publish.interval=10, -Xms128m, -Xmx256m, -Djava.security.egd=file:/dev/./urandom, -Dagent.plugins.md5=47383b4ad0fb2c52313aad5e77b18a03, -Dagent.binary.md5=5e51a7cb2e55fa232e8a1d25e6e2aba7, -Dagent.launcher.md5=9c67d216e09d2b82a17c8fcfd3bd48c5, -Dagent.tfs.md5=fa884c713ea0dc9b1430dc551155124f]
[Tue Aug 28 11:05:41 UTC 2018]          JVM properties: {java.runtime.name=OpenJDK Runtime Environment, sun.boot.library.path=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/amd64, java.vm.version=25.181-b13, java.vm.vendor=Oracle Corporation, java.vendor.url=http://java.oracle.com/, path.separator=:, java.vm.name=OpenJDK 64-Bit Server VM, file.encoding.pkg=sun.io, user.country=US, sun.java.launcher=SUN_STANDARD, sun.os.patch.level=unknown, java.vm.specification.name=Java Virtual Machine Specification, user.dir=/go, java.runtime.version=1.8.0_181-b13, java.awt.graphicsenv=sun.awt.X11GraphicsEnvironment, java.endorsed.dirs=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/endorsed, os.arch=amd64, agent.binary.md5=5e51a7cb2e55fa232e8a1d25e6e2aba7, java.io.tmpdir=/tmp, line.separator=
, java.vm.specification.vendor=Oracle Corporation, os.name=Linux, sun.jnu.encoding=ANSI_X3.4-1968, agent.plugins.md5=47383b4ad0fb2c52313aad5e77b18a03, java.library.path=/usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib, java.specification.name=Java Platform API Specification, java.class.version=52.0, sun.management.compiler=HotSpot 64-Bit Tiered Compilers, os.version=4.14.56+, user.home=/go, user.timezone=UTC, agent.launcher.md5=9c67d216e09d2b82a17c8fcfd3bd48c5, java.awt.printerjob=sun.print.PSPrinterJob, file.encoding=ANSI_X3.4-1968, java.specification.version=1.8, java.class.path=agent.jar:/usr/local/bin/jmx_prometheus_javaagent.jar, user.name=go, cruise.console.publish.interval=10, java.vm.specification.version=1.8, sun.java.command=agent.jar -serverUrl https://goqa.atcloud.io/go -sslVerificationMode FULL, java.home=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre, sun.arch.data.model=64, user.language=en, java.specification.vendor=Oracle Corporation, awt.toolkit=sun.awt.X11.XToolkit, java.vm.info=mixed mode, java.version=1.8.0_181, java.ext.dirs=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/ext:/usr/java/packages/lib/ext, sun.boot.class.path=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/resources.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/rt.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/sunrsasign.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/jsse.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/jce.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/charsets.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/lib/jfr.jar:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-3.b13.el7_5.x86_64/jre/classes, java.vendor=Oracle Corporation, java.security.egd=file:/dev/./urandom, file.separator=/, java.vendor.url.bug=http://bugreport.sun.com/bugreport/, sun.io.unicode.encoding=UnicodeLittle, sun.cpu.endian=little, agent.tfs.md5=fa884c713ea0dc9b1430dc551155124f, sun.cpu.isalist=}
[Tue Aug 28 11:05:41 UTC 2018]   Environment Variables: {PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, DOCKER_HOST=tcp://127.0.0.1:2375, KUBERNETES_PORT=tcp://10.192.104.1:443, JAVA_HOME=/usr/lib/jvm/java-openjdk, HEAP_DUMP_DIR=/atcloud/heapdumps, TERM=xterm-256color, KUBERNETES_SERVICE_HOST=10.192.104.1, GOOGLE_APPLICATION_CREDENTIALS=/etc/gcloud/gcloud_json, APP_ADMIN_PORT=9080, PWD=/go, GOCD_AGENT_PULL_BASE=false, KUBERNETES_PORT_443_TCP=tcp://10.192.104.1:443, APP_PATH=/usr/local/autotrader/app, KUBERNETES_PORT_443_TCP_ADDR=10.192.104.1, APP_DIR=/usr/local/autotrader/app, KUBERNETES_PORT_443_TCP_PROTO=tcp, KUBERNETES_SERVICE_PORT=443, JMX_METRICS_PORT=9090, GO_API_PASSWORD=<REDACTED>, GO_SERVER_API_URL=https://goqa.atcloud.io/go/api, HOSTNAME=gocd-admin-agent-5e226ebe-55c8-4142-a503-a04e7551222c, KUBERNETES_PORT_443_TCP_PORT=443, KUBERNETES_SERVICE_PORT_HTTPS=443, APP_PORT=8080, TEST_ENV="true", HOME=/go, SHLVL=0}
Could not find file `config/agent-logback.xml'. Attempting to load from classpath.
Using classpath resource `jar:onejar:lib/agent-18.6.0-6883-classes.jar!/config/agent-logback.xml'.
```

Another thing to note is I had to switch from using the `:8154` server address (Which uses GoCD's self signed cert) which worked for the normal gocd docker agent, to using our `:443` server address which uses a public cert

**Comment by Evesy** [comment]
Thanks @ketan. I've tried disabling SSL verification & using port 8154 however it doesn't appear that `GO_EA_SSL_NO_VERIFY` is being honoured by the bootstrapper:

```
2018-08-28T11:39:06.475Z+00:00 [Bootstrap] [DEBU] Getting checksums from https://goqa.atcloud.io:8154/go/admin/latest-agent.status
2018-08-28T11:39:06.681Z+00:00 [Bootstrap] [CRIT] Could not fetch URL https://goqa.atcloud.io:8154/go/admin/latest-agent.status. Head https://goqa.atcloud.io:8154/go/admin/latest-agent.status: x509: certificate is valid for qaapp246.qa.dc1.tradermedia.net, not goqa.atcloud.io
```

It might be worth saying that our current elastic agents (That use https://github.com/gocd/docker-gocd-agent-centos-7) can connect/register fine, without setting `GO_EA_SSL_NO_VERIFY` to false

**Comment by ketan** [comment]
Seems like a bug — you'd need to set `GO_EA_SSL_NO_VERIFY` to `false` instead of `true`. I'll fix this in the next release.

**Comment by ketan** [comment]
I ran with this shell script BTW:

```shell
#!/bin/bash

docker run -d -p8153:8153 -p8154:8154 gocd/gocd-server:v18.7.0

(cat <<-DockerFile
FROM alpine:3.7
ENV LANG=en_US.utf8

ARG UID=1000
ARG GID=1000

RUN addgroup -g \${GID} go && \ 
  adduser -D -u \${UID} -s /bin/bash -G go go && \
  apk --no-cache upgrade && \
  apk add --no-cache openjdk8-jre-base git bash curl && \
  curl --fail --silent --location https://github.com/ketan/gocd-golang-bootstrapper/releases/download/2.1/go-bootstrapper-2.1.linux.amd64 > /go-agent && \
  chmod 755 /go-agent

RUN mkdir -p /go && \
  chown go:go /go
USER go
CMD /go-agent
DockerFile
) | docker build -t custom-agent -

docker run -it --rm \
  -eGO_EA_AUTO_REGISTER_KEY=139d1922-e9ab-4687-afdf-bb01f56a4b63 \
  -eGO_EA_AUTO_REGISTER_ELASTIC_AGENT_ID=foo \
  -eGO_EA_SERVER_URL=https://192.168.0.5:8154/go \
  -eGO_EA_AUTO_REGISTER_ELASTIC_PLUGIN_ID=foo \
  -eGO_EA_SSL_NO_VERIFY=false \
  custom-agent
```

**Comment by ketan** [comment]
Hi — I've released version 2.2 of the bootstrapper that should fix the issue with `GO_EA_SSL_NO_VERIFY` being interpreted incorrectly. Were you able to get this working as expected?

**Comment by arvindsv** [comment]
Closing. No activity.
